File: lingxia-sdk/android/lingxia/src/main/java/com/lingxia/app/LxLog.kt
package com.lingxia.app

/**
 * Routes SDK-native logs into the LingXia Rust log pipeline.
 *
 * Unlike [android.util.Log], records emitted here flow through the same pipeline
 * as Rust logs: they reach logcat *and* the dev-server stream, so they show up
 * in `lxdev logs` tagged with the originating [appId]/[path].
 *
 * **Forward errors and important warnings only.** Records here are also buffered
 * for cloud upload / crash diagnosis, so routing routine info/debug dilutes that
 * bounded buffer (evicting the errors it's meant to keep) and pays an FFI
 * crossing per call. Keep lifecycle/info/debug and high-frequency traces on
 * `android.util.Log`; send through `LxLog` the diagnostics you'd want in an
 * uploaded log bundle. On a hot path, guard with `LxLog.isEnabled(level)`.
 *
 * Method names and primary argument order follow `android.util.Log`; optional
 * appId/path metadata routes records to the owning lxapp in dev logs.
 */
internal object LxLog {
    // Mirrors the Rust FFI level contract (see `logging::forward_host_log`).
    private const val VERBOSE = 0
    private const val DEBUG = 1
    private const val INFO = 2
    private const val WARN = 3
    private const val ERROR = 4

    /** Whether a log at [level] (0=verbose … 4=error) would be recorded. Guard an
     *  expensive hot-path log with this to skip building the message. */
    fun isEnabled(level: Int): Boolean = NativeApi.hostLogEnabled(level)

    fun v(tag: String, message: String, appId: String = "", path: String = "") =
        NativeApi.forwardHostLog(VERBOSE, tag, appId, path, message)

    fun d(tag: String, message: String, appId: String = "", path: String = "") =
        NativeApi.forwardHostLog(DEBUG, tag, appId, path, message)

    fun i(tag: String, message: String, appId: String = "", path: String = "") =
        NativeApi.forwardHostLog(INFO, tag, appId, path, message)

    fun w(tag: String, message: String, tr: Throwable? = null, appId: String = "", path: String = "") =
        NativeApi.forwardHostLog(WARN, tag, appId, path, message.withThrowable(tr))

    fun e(tag: String, message: String, tr: Throwable? = null, appId: String = "", path: String = "") =
        NativeApi.forwardHostLog(ERROR, tag, appId, path, message.withThrowable(tr))

    private fun String.withThrowable(tr: Throwable?): String =
        if (tr == null) this else "$this\n${tr.stackTraceToString()}"
}

File: lingxia-sdk/apple/Sources/Support/LXLog.swift
import CLingXiaRustAPI

/// Routes SDK-native logs into the LingXia Rust log pipeline.
///
/// Unlike `os_log`, records emitted here flow through the same pipeline as Rust
/// logs: they reach the platform sink *and* the dev-server stream, so they show
/// up in `lxdev logs` tagged with the originating `appId`/`path`.
///
/// **Forward errors and important warnings only.** Records here are also
/// buffered for cloud upload / crash diagnosis, so routing routine info/debug
/// dilutes that bounded buffer (evicting the errors it's meant to keep) and
/// pays an FFI crossing + eager message build per call. Keep lifecycle/info/
/// debug and high-frequency traces on `os_log`; send through `LXLog` the
/// diagnostics you'd want in an uploaded log bundle. On a genuinely hot path
/// guard the call with ``isEnabled(_:)``.
enum LXLog {
    /// Mirrors the Rust FFI level contract (see `logging::forward_host_log`).
    enum Level: Int32 {
        case verbose = 0
        case debug = 1
        case info = 2
        case warn = 3
        case error = 4
    }

    /// Whether a log at `level` would be recorded at the current threshold.
    /// Guard an expensive hot-path log with this to skip building the message:
    /// `if LXLog.isEnabled(.debug) { LXLog.debug("…\(costly())…", category: "X") }`.
    static func isEnabled(_ level: Level) -> Bool {
        hostLogEnabled(level.rawValue)
    }

    /// Forward a log entry into the Rust pipeline.
    ///
    /// `message` is a plain `String`, not an autoclosure: `forwardHostLog`
    /// evaluates and dispatches unconditionally, so a deferred closure would
    /// buy nothing and only imply a laziness that doesn't exist.
    /// - Parameters:
    ///   - message: Fully-formatted message.
    ///   - category: Subsystem/category label, surfaced as the log target.
    ///   - appId: Owning lxapp id, when known. Empty for host-global logs.
    ///   - path: Page path within the lxapp, when known.
    @discardableResult
    static func log(
        _ level: Level,
        _ message: String,
        category: String,
        appId: String = "",
        path: String = ""
    ) -> Bool {
        forwardHostLog(level.rawValue, category, appId, path, message)
    }

    @discardableResult
    static func verbose(_ message: String, category: String, appId: String = "", path: String = "") -> Bool {
        log(.verbose, message, category: category, appId: appId, path: path)
    }

    @discardableResult
    static func debug(_ message: String, category: String, appId: String = "", path: String = "") -> Bool {
        log(.debug, message, category: category, appId: appId, path: path)
    }

    @discardableResult
    static func info(_ message: String, category: String, appId: String = "", path: String = "") -> Bool {
        log(.info, message, category: category, appId: appId, path: path)
    }

    /// `error:` mirrors Android/Harmony `w(tag, msg, tr)`: when present its
    /// description is appended so caught errors carry context across platforms.
    @discardableResult
    static func warn(_ message: String, category: String, appId: String = "", path: String = "", error: Error? = nil) -> Bool {
        log(.warn, Self.appending(error, to: message), category: category, appId: appId, path: path)
    }

    @discardableResult
    static func error(_ message: String, category: String, appId: String = "", path: String = "", error: Error? = nil) -> Bool {
        log(.error, Self.appending(error, to: message), category: category, appId: appId, path: path)
    }

    private static func appending(_ error: Error?, to message: String) -> String {
        guard let error else { return message }
        return "\(message)\n\(error)"
    }
}
